jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/service/WildFlyCustomJtaPlatform.java

```diff
@@ -24,37 +24,69 @@
-                locateTransactionSynchronizationRegistry();
+    // The 'transactionSynchronizationRegistry' used by JPA container managed applications,
+    // is reset every time the Transaction Manager service is restarted,
+    // as (application deployment) JPA persistence unit service depends on the TM service.
+    // For this reason, the static 'transactionSynchronizationRegistry' can be updated.
+    // Note that Hibernate native applications currently have to be (manually) restarted when the TM
+    // service restarts, as native applications do not have WildFly service dependencies set for them.
+    private static volatile TransactionSynchronizationRegistry transactionSynchronizationRegistry;
 
-        return new JtaSynchronizationStrategy() {
+    // JtaSynchronizationStrategy
+    @Override
+    public void registerSynchronization(Synchronization synchronization) {
+        transactionSynchronizationRegistry.registerInterposedSynchronization(synchronization);
+    }
 
+    // JtaSynchronizationStrategy
+    @Override
+    public boolean canRegisterSynchronization() {
+        return transactionSynchronizationRegistry.getTransactionStatus() == Status.STATUS_ACTIVE;
+    }
 
-            @Override
-            public void registerSynchronization(Synchronization synchronization) {
-                transactionSynchronizationRegistry.registerInterposedSynchronization(synchronization);
-            }
+    @Override
+    protected JtaSynchronizationStrategy getSynchronizationStrategy() {
+        return this;
+    }
 
-            @Override
-            public boolean canRegisterSynchronization() {
-                return transactionSynchronizationRegistry.getTransactionStatus() == Status.STATUS_ACTIVE;
+    private TransactionSynchronizationRegistry locateTransactionSynchronizationRegistry() {
+        if (transactionSynchronizationRegistry == null) {
+            synchronized (WildFlyCustomJtaPlatform.class) {
+                if (transactionSynchronizationRegistry == null) {
+                    transactionSynchronizationRegistry = (TransactionSynchronizationRegistry) jndiService().locate(TSR_NAME);
+                }
             }
-        };
-
+        }
+        return transactionSynchronizationRegistry;
     }
 
-    private TransactionSynchronizationRegistry locateTransactionSynchronizationRegistry() {
-        return (TransactionSynchronizationRegistry) jndiService().locate(TSR_NAME);
+    /**
+     * Hibernate native applications cannot know when the TransactionManager + TransactionSynchronizationRegistry
+     * services are stopped but JPA container managed applications can and will call setTransactionSynchronizationRegistry
+     * with the new (global) TransactionSynchronizationRegistry to use.
+     *
+     * @param tsr
+     */
+    public static void setTransactionSynchronizationRegistry(TransactionSynchronizationRegistry tsr) {
+
+        if (tsr != transactionSynchronizationRegistry) {
```

Comment:
Do we want to be checking the object instance itself or should it be using `equals()`?

Reply:
If we switch to calling equals(), that requires all possible implementations of TransactionSynchronizationRegistry to handle the object equivalence checking for us, currently, we are doing a check for the same object instance (should always be correct).  The only concern that I have is if every call to setTransactionSynchronizationRegistry was changed to pass a new TSR wrapper, instead of the underlying TSR object, which would definitely be wrong for this code.